utils/xsendkeys.8
.TH XSENDKEYS 8 "March 2, 2005"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
xsendkeys \- send a named key combination to X using strings
.SH SYNOPSIS
.B xsendkeys keystring 

.SH "DESCRIPTION"
xsendkeys is a simple program which allows you to send a key or key combination to X. For example:
.B  xsendkeys "a"
will send the letter a. If you wanted to send a capital a 'A' you would have to send:
.B  xsendkeys "Shift_L+a"
.PP
.B  xsendkeys "Shift_L+Right" 
will send the key combination shift+right arrow. The names of the keys must be the exact names X would expect (use xev to see keynames when in doubt). Unlike xsendkeycode xsendkeys automatically sends corresponding keydown and keyup events. This helper application is not used from lineakd(1) but is included to allow scripts to send keys to X. 

.SH SEE ALSO
evtest(8),
send_to_keyboard(8),
xsendkeycode(8),
lineakd(1).
.SH AUTHOR
.B xsendkeys
was written by Sheldon Lee-Wen <leewsb@hotmail.com>.


ChangeLog
June 9 2006     Sheldon Lee-Wen
		Fix bug in Makefile.am that causes duplicate DESTDIR's when
		DESTDIR is specified. This bug is triggered under newer
		autotools when creating RPMS.

		Fix bug that caused a segfault after a lineakd -r was issued
		and lineakd attempted to run a command that was not a plugin.

		Fix a bug that caused lineak to revert to the internal display
		plugin when a MappingNotify event is recieved.
		
May 24 2006     Sheldon Lee-Wen
		Released 0.9. Adds more supported keyboards. Fixes bugs.
		Provides API required by KlineakConfig.

March 13 2006   Sheldon Lee-Wen
                Many bug fixes within the lineak library and API. Some API
		changes to support KlineakConfig as well as to fix borked
		APIs. Many keyboards added.

July 22 2005    Sheldon Lee-Wen
		lineakd 0.8.4-rc1. We now have support for RAWCOMMAND[SETKEYCODES] = 
		directives in the lineakkb.def file. If setkeycodes is suid root
		then lineakd will run setkeycodes for us. The reason for this is two fold.
		1) Convienience for the user. 2) A documented place to keep
		scancodes for devices which are not recognized. Once we start
		getting these, I am going to look at contributing them to the kernel to
		fix the problem where is should be fixed.

May 26 2005     Sheldon Lee-Wen
		lineakd 0.8.4 development Fixed some bugs, introduced helper applications 
                xsendkeys and xsendkey.

February 25 2005 Sheldon Lee-Wen
                Several bugs have been fixed. Most seriously the AMD64 bugs. 
		The code is now much more portable. I've also begun
		implementing support for an external helper application, but
		this has to be run as suid root, and can send commands
		directly to the keyboard, so it's a little dangerous. The
		helper command sends commands directly to the keyboard, and
		the keyboard definition can be configured to support this with
		a RAWCOMMANDS[INITIALIZE] = <commands> entry. It's just in the
		initial stages of support, so may not work for everything.

		We've also updated the lineakkd.def file with more keyboards.

November  8 2004 Sheldon Lee-Wen
                Finally fixed a long standing bug with the PluginManager class 
		that would segfault if you tried to unload plugins. Turns out
		it was in the initializeDisplay method of the class. I added
		many diagnostics functions in order to track this down.

November  2 2004 Sheldon Lee-Wen
                Posting lineakd 0.8.2. Nothing has really changed except some
		compile time fixes for SUSE 9.1, which has to do with which
		headers are included where, and in which order.

October  23 2004 Sheldon Lee-Wen 
                Lineak 0.8.1 is ready to go out. In this release we have bug
		fixes, and versioning in the plugin structures so that we
		don't load an inappropriate version. I'm also going to begin 
		converting the plugin structure to use strings instead of 
		char*. Seeing that this is C++ anyway.
		   The big new feature for 0.8.1 is the inclusion of a
		   Media-Detect macro, but that is in the defaultplugins
		   package.

September 4 2004 Sheldon Lee-Wen
                I'm preparing to release 0.8 final. In doing that I've added a
                last minute feature for custom on screen display messages.
                In the config file you can now do the following:
                 [On Screen Message] key+modifier = command
                To have the "On Screen Message" appear on the screen when the
                key in question has been pressed.

April 24 2004 Sheldon Lee-Wen
                I'm about to release the third beta of lineakd 0.8
                We now support buttons and keys. Currently to bind to a button, you have to place a custom
                lineakkb.def file in your home directory. In the definition for your keyboard, you must add
                a section for buttons like in the following example: 
                [LTCDP]
                   brandname = "Logitech"
                   modelname = "Cordless Desktop Pro"
                   [KEYS]
                      Sleep                 = 223
                      Internet              = 178
                      Mail                  = 236
                   [END KEYS]
                   [BUTTONS]
                      Thumb                 = 2
                   [END BUTTONS]
               [END LTCDP]
           
               We also now support a variable number of modifiers to a key. This only applies to non-toggleable keys
               however, so a key such as Play|Pause will not work with modifiers if you use it as a toggleable key,
               by putting entries like this:
                  Play = something
                  Pause = something else
               in your config file. However, if you use the key as a non-toggleable key, i.e. like this:
                  Play|Pause = something
               then it becomes possible to use modifiers.

               Currently we support the following modifiers:
                  control 
                  alt 
                  shift 
                  mod2 
                  mod3 
                  mod4 
                  mod5
               To use a modifier, we do something like this in the config file:
                  Sleep+control = something
                  Sleep+alt = something different
                  Sleep+shift = something more different
                  Sleep = sleepiness
               etc.

               

February 2 2004 Sheldon Lee-Wen
		This is the second public beta release of lineakd 0.8
                The X handling code has been ripped out and recoded in a much simplified manner.
                However, it needs a lot of testing to determine if it will work properly.
                Also, on exit and restarts, we are not closing and unloading the plugins from the plugin manager. 
                I still have to track down a bug wrt plugin unloading, which means that a restart of

utils/xsendkeys.cpp
/* xsendkey.c ************************************************/
/*
 * This file is subject to the terms and conditions of the GNU General Public
 * License.  See the file COPYING in the main directory of this archive for
 * more details.
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/extensions/XTest.h>
#include <lineak/lineak_core_functions.h>
#include <lineak/lineak_util_functions.h>
#include <lineak/lobject.h>

#include <vector>
#include <iostream>
#include <stdlib.h>

using namespace std;
using namespace lineak_core_functions;
using namespace lineak_util_functions;

char *ProgramName;

Display *dpy;
int scr;
Window win;
unsigned int width, height;

int main(int argc, char **argv)
{
    Window ret_win;
    int x, y;
    unsigned int border_width, depth;
    int keycode;
    string argument;

    ProgramName = argv[0];
    if (argc < 1)
    	return 1;

    argument = string(argv[1]);
    
    dpy = XOpenDisplay("");
    if (!dpy) {
	fprintf(stderr, "%s: Cannot connect to display ...\n", ProgramName);
	return 1;
    }
    scr = DefaultScreen(dpy);
    win = RootWindow(dpy, scr);
    XGetGeometry(dpy, win, &ret_win, &x, &y, &width, &height, &border_width, &depth);
    /////////////////////////////////
    // lineak stuff
    vector<string> args;
    args.push_back(argument);
    vector<string> elements;
    vector<int>codes;
    KeySym mykeysym;
    KeyCode xkeycode;
    
    int index = 0;
    string parsed = "";
    string symname = lineak_util_functions::strip_space(args[0]);
    string smodifiers = "";
    while (symname.find('+') != string::npos) {
	    index = symname.find('+');
	    smodifiers = symname.substr(index+1,symname.size()-(index+1));
	    parsed = symname.substr(0,index);
	    symname = smodifiers;
	    if ( parsed == "control" || parsed == "Control" || parsed == "CONTROL" )
		    parsed = "Control_L";
	    if ( parsed == "shift" || parsed == "Shift" || parsed == "SHIFT" )
		    parsed = "Shift_L";
	    if ( parsed == "mod1" || parsed == "alt" || parsed == "Alt" || parsed == "ALT" )
		    parsed = "Alt_L";
            //cout << "element = " << parsed << endl;
	    elements.push_back(parsed);
	    //cout << "remainder = " << smodifiers << endl;
    }
    elements.push_back(smodifiers);
    //inspect_vector(elements);
    
    for (vector<string>::const_iterator it = elements.begin(); it != elements.end(); it++) {
        mykeysym = XStringToKeysym(it->c_str());
        xkeycode = XKeysymToKeycode(dpy, mykeysym);	
	//cout << *it << " has code of " << (int)xkeycode << endl; 
	codes.push_back((int)xkeycode);
    }
    /*
    cout << "vector has " << codes.size() << " elements: { ";
    for (vector<int>::const_iterator it = codes.begin(); it != codes.end(); it++)
		                    cout << *it << " ";
    cout << "} " << endl;
    */
    for (vector<int>::const_iterator it = codes.begin(); it != codes.end(); it++) {
	    keycode = *it;
	    XTestFakeKeyEvent(dpy, keycode, 1, 0);
	    //cout << "sending key down: " << keycode << endl;
	    XSync(dpy, 1);
    }
    vector<int>::const_iterator it = codes.end();
    
    do {
	    it--;
	    keycode = *it;
	    XTestFakeKeyEvent(dpy, keycode, 0, 0);
	    //cout << "sending key up: " << keycode << endl;
	    XSync(dpy, 1);
    }
    while (it != codes.begin());
	    
    return 0;
}
